src/wrek_vert_t.erl

```diff
@@ -0,0 +1,301 @@
+
+-type t() :: #t{}.
+
+-export_type([
+    t/0
+]).
+
+
+-spec new() -> t().
+
+new() ->
+    #t{}.
+
+
+-spec cancel(t()) -> t().
+
+cancel(Vert = #t{}) ->
+    set_status(Vert, cancelled).
+
+
+-spec fail(t(), reason_t()) -> t().
+
+fail(Vert = #t{}, Reason) ->
+    Vert2 = set_reason(Vert, Reason),
+    set_status(Vert2, failed).
+
+
+-spec from_defn(map()) -> {ok, t()} | {error, any()}.
+
+from_defn(Map0) when is_map(Map0) ->
+    Res0 = #t{},
+
+    MandatoryFields = [
+        {module, fun set_module/2},
+        {args, fun set_args/2},
+        {deps, fun set_deps/2}
+    ],
+
+    OptionalFields = [
+        {name, fun set_name/2},
+        {timeout, fun set_timeout/2}
+    ],
+
+    case load_mandatory(Res0, Map0, MandatoryFields) of
+        {error, _} = Err ->
+            Err;
+        {ok, {Res1, Map1}} ->
+            {ok, {Res2, Map2}} = load_optional(Res1, Map1, OptionalFields),
+            Res3 = set_kv(Res2, Map2),
+            {ok, Res3}
+    end;
+
+from_defn(_) ->
+    {error, not_a_map}.
+
+
+-spec has_succeeded(t()) -> boolean().
+
+has_succeeded(Vert = #t{}) ->
+    case status(Vert) of
+        done ->
+            true;
+        _ ->
+            false
+    end.
+
+
+-spec is_finished(t()) -> boolean().
+
+is_finished(Vert = #t{}) ->
+    case status(Vert) of
+        done ->
+            true;
+        failed ->
+            true;
+        cancelled ->
+            true;
+        _ ->
+            false
+    end.
+
+
+-spec succeed(t(), map()) -> t().
+
+succeed(Vert = #t{}, Result) ->
+    Vert2 = set_kv(Vert, Result),
+    set_status(Vert2, done).
+
+
+-spec args(t()) -> args_t().
+
+args(#t{args = Args}) ->
+    Args.
+
+
+-spec deps(t()) -> deps_t().
+
+deps(#t{deps = Deps}) ->
+    Deps.
+
+
+-spec dir(t()) -> dir_t().
+
+dir(#t{dir = Dir}) ->
+    Dir.
+
+
+-spec id(t()) -> id_t().
+
+id(#t{id = Id}) ->
+    Id.
+
+
+-spec kv(t()) -> kv_t().
+
+kv(#t{kv = Kv}) ->
+    Kv.
+
+
+-spec module(t()) -> module_t().
+
+module(#t{module = Module}) ->
+    Module.
+
+
+-spec name(t()) -> name_t().
+
+name(#t{name = Name}) ->
+    Name.
+
+
+-spec reason(t()) -> reason_t().
+
+reason(#t{reason = Reason}) ->
+    Reason.
+
+
+-spec status(t()) -> status_t().
+
+status(#t{status = Status}) ->
+    Status.
+
+
+-spec timeout(t()) -> timeout_t().
+
+timeout(#t{timeout = Timeout}) ->
+    Timeout.
+
+
+-spec set_args(t(), args_t()) -> t().
+
+set_args(T = #t{}, Args) ->
+    T#t{args = Args}.
+
+
+-spec set_deps(t(), deps_t()) -> t().
+
+set_deps(T = #t{}, Deps) ->
+    T#t{deps = Deps}.
+
+
+-spec set_dir(t(), dir_t()) -> t().
+
+set_dir(T = #t{}, Dir) ->
+    T#t{dir = Dir}.
+
+
+-spec set_id(t(), id_t()) -> t().
+
+set_id(T = #t{}, Id) ->
+    T#t{id = Id}.
+
+
+-spec set_kv(t(), kv_t()) -> t().
+
+set_kv(T = #t{}, Kv) ->
+    T#t{kv = Kv}.
+
+
+-spec set_module(t(), module_t()) -> t().
+
+set_module(T = #t{}, Module) ->
+    T#t{module = Module}.
+
+
+-spec set_name(t(), name_t()) -> t().
+
+set_name(T = #t{}, Name) ->
```

Comment:
More of a question but is there a reason to have a setter and getter for everything or is it a style you prefer? This feels very java~y.

Reply:
One reason for all the different setters is that most of the fields in the record are optional, and are kind of set one-at-a-time in `from_defn/1` and `load/4`. You're right in that it feels very OO, but, on the bright side, the 'setters' are actually returning copies of the input, rather than mutating the input directly.

Reply:
Ah yea I had no doubt they weren't actually mutating anything. I guess it just seemed like useless abstraction.